File: tests/templates/TodayExtension/TodayExtensionMain.cs
using System;

using NotificationCenter;
using Foundation;
using UIKit;

using MonoTouch.NUnit.UI;
using NUnit.Framework;
using NUnit.Framework.Internal;
using NUnit.Framework.Internal.Filters;

[Register ("TodayViewController")]
public partial class TodayViewController : UIViewController, INCWidgetProviding
{
	ConsoleRunner runner;

	protected TodayViewController (IntPtr handle) : base (handle) {}

	[Export ("widgetPerformUpdateWithCompletionHandler:")]
	public void WidgetPerformUpdate (Action<NCUpdateResult> completionHandler)
	{
		runner = new ConsoleRunner ();
		runner.Add (System.Reflection.Assembly.GetExecutingAssembly ());
		runner.Filter = new NotFilter (new CategoryExpression ("MobileNotWorking,NotOnMac,NotWorking,ValueAdd,CAS,InetAccess,NotWorkingLinqInterpreter").Filter);
		System.Threading.ThreadPool.QueueUserWorkItem ((v) =>
		{
			runner.LoadSync ();
			BeginInvokeOnMainThread (() =>
			{
				runner.AutoStart = true; // There's no UI for today extensions (yet), so unless we autostart the tests, there's no way to run them when running from the IDE.
				runner.AutoRun ();
			});
		});

		completionHandler (NCUpdateResult.NewData);
	}
}

File: tests/templates/TodayExtension/TodayExtensionMain.fs
namespace todayextensiontests

open System
open System.Threading

open NotificationCenter
open Foundation
open UIKit

open MonoTouch.NUnit.UI
open NUnit.Framework
open NUnit.Framework.Internal
open NUnit.Framework.Internal.Filters

[<Register ("TodayViewController")>]
type TodayViewController() =
    inherit UIViewController()

    let mutable runner = Unchecked.defaultof<ConsoleRunner>

    [<Export ("widgetPerformUpdateWithCompletionHandler:")>]
    member this.WidgetPerformUpdate (completionHandler: Action<NCUpdateResult>) =
        runner <- new ConsoleRunner ()
        let ce = new CategoryExpression ("MobileNotWorking,NotOnMac,NotWorking,ValueAdd,CAS,InetAccess,NotWorkingLinqInterpreter")
        runner.Filter <- new NotFilter (ce.Filter)
        let tp = this.GetType ()
        runner.Add (tp.Assembly)
        ThreadPool.QueueUserWorkItem (fun v ->
            runner.LoadSync ()
            this.BeginInvokeOnMainThread (fun x ->
                runner.AutoStart <- true
                runner.AutoRun ()
            )
        )
        |> ignore
        completionHandler.Invoke (NCUpdateResult.NewData)

    interface INCWidgetProviding
